model-optimizer/extensions/ops/If.py

```diff
@@ -0,0 +1,321 @@
+        :return: None
+        """
+        subgraph = if_node.then_graph if condition else if_node.else_graph
+        for node in subgraph.get_op_nodes():
+            if node.has('input_id'):
+                assert node.soft_get('type') == 'Parameter'
+                input_port_id = node['input_id']
+                input_shape = if_node.in_port(input_port_id).data.get_shape()
+                if node.soft_get('shape', None) is None:
+                    node['shape'] = None
+                node.shape = input_shape.copy()
+                log.debug('Updated shape for the body node with name "{}" with value {}'
+                          .format(node.soft_get('name', node.soft_get('id')), node.shape))
+
+    @staticmethod
+    def update_if_output_ports_shape(if_node: Node):
+        """
+        Update shape and values for If output ports.
+
+        :param if_node: The If node to update output ports and shapes
+        :return: None
+        """
+
+        then_outputs = [node for node in if_node.then_graph.get_op_nodes() if node.has('output_id')]
+        else_outputs = [node for node in if_node.else_graph.get_op_nodes() if node.has('output_id')]
+        outputs_mapping = {}
+        outputs_number = len(if_node.out_ports())
+
+        if outputs_number == 0 and len(if_node.out_ports(control_flow=True)) != 0:
+            # Some models have if with control flow outputs.
+            # These shape inference for such ifs
+            # TODO: need to rethink and redo support for control flow edges in if operation
+            for node in if_node.out_nodes(control_flow=True).values():
+                node.shape = int64_array([])
+            return
+
+        for port_id in if_node.out_ports().keys():
+            outputs_mapping[port_id] = {}
+        port_ids = outputs_mapping.keys()
+
+        # variables then_contains_fake_outputs/else_contains_fake_outputs contains True value
+        # if all outputs from then_body/else_body have shape [0]. It means then_body/else_body does not return data
+        # and further shape_inference for this branch is not possible.
+        # TODO: exclude support fake_outputs from this code when we will support shape_inference with empty tensors
+        then_contains_fake_outputs = True
+        else_contains_fake_outputs = True
+
+        for then_output_node in then_outputs:
+            assert then_output_node.soft_get('type') == 'Result'
+            port_id = then_output_node['output_id']
+            assert port_id in port_ids, 'Incorrect mapping then_graph outputs with {0} outputs! ' \
+                                        'Can\'t find port with ID {1} in If operation.' \
+                .format(then_output_node.name, port_id)
+            outputs_mapping[port_id]['then_graph'] = then_output_node
+            then_shape = then_output_node.in_port(0).data.get_shape()
+            then_contains_fake_outputs = then_contains_fake_outputs and np.any(then_shape == 0)
```

Comment:
Let's extract these lines into a function because they are equal to lines 161-169.

Reply:
nit

Reply:
done